rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md

```diff
@@ -1 +1,157 @@
+* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
+* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.
+
+| `hash_type` | matches by       | VM version |
+| ----------- | ---------------- | ---------- |
+| 0           | data hash        | 0          |
+| 1           | type script hash | 1          |
+| 2           | data hash        | 1          |
+
+The VM version 0 deployed with the Lina and Aggron genesis block is version 0.
+
+The VM version 1 will be deployed along ckb2021. The [rfc33] highlights the changes in this version, and [rfc34] lists the new syscalls only available since version 1.
+
+[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md
+[rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md
+
+The transaction is considered invalid if any `hash_type` is not in the allowed values 0, 1, and 2.
+
+See more information about code locating using `hash_type` in [rfc22].
+
+[rfc22]: ../0022-transaction-structure/0022-transaction-structure.md
+
+Because the VM selection algorithm depends on which epoch the transaction belongs to, it is not deterministic for transactions still in the memory pool. The CKB node must run two versions of transaction relay protocols, one for the current CKB version C and another for the next fork version N.
+
+* Before the fork is activated, CKB node must relay transactions via relay protocol C and must drop all messages received via protocol N.
+* After the fork is activated, CKB node must relay transactions via relay protocol N and must drop all messages received via protocol C.
+
+The relay protocol C will be dropped after the fork succeeds. See [rfc35] for details.
+
+[rfc35]: ../0035-ckb2021-p2p-protocol-upgrade/0035-ckb2021-p2p-protocol-upgrade.md
+
+Attention that, [rfc33], [rfc34], and [rfc35] must be activated together with this one starting at the same epoch.
+
+When a new block is appended to the chain and the fork is activated, or a block is rolled back and the fork is deactivated, the CKB node must rerun the verification on all the transactions in the pool.
+
+In [nervosnetwork/ckb](https://github.com/nervosnetwork/ckb), the `hash_type` is returned in the JSON RPC as an enum. Now it has three allowed values:
+
+* 0: "data"
+* 1: "type"
+* 2: "data1"
+
+## Rationale
+
+There are many other solutions to select VM versions. The current solution is the result of discussion and trade-off. Following are some example alternatives:
+
+* Always use the latest VM version. The users have no options to freeze the VM versions used in their transactions.
+* Depend on the script code cell epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The problem of this solution is that anyone can re-deploy the cell and construct the transaction using the new code cell to choose VM versions.
```

Comment:
```suggestion
Depend on the script code cell deployed epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The drawback of this proposal is that anyone can re-deploy the script code after the fork, then the transaction using the newly deployed code cell will choose new VM versions. This may be unexpected behavior.
```